chore(README): update gateway readme (#612)
ProjectLibertyLabs/gateway: README.md

# Gateway
Gateway is a suite of microservices designed to simplify interactions with the Frequency blockchain, bridging the gap between Web2 and Web3 development.

<!-- TABLE OF CONTENTS -->

# 📗 Table of Contents

- [📖 About the Project](#about-project)
- [🚀 Live OpenAPI Docs](#-live-docs)
- [🔍 Architecture Map](#-arch-maps)
- [🔍 Gateway Microservices](#gateway-microservices)
- [💻 Getting Started](#getting-started)
  - [🚀 Quick Start Guide](#quick-start-guide)
  - [💻 Microservice Start Guide](#microservices-start-guide)

> - [Deployment](#deployment)

Reviewer: Maybe different icons and add one for deployment.

# 📖 Gateway Services <a name="about-project"></a>

> Gateway is a collection of services designed to simplify interactions with the Frequency blockchain. It bridges the gap between Web2 and Web3, allowing developers to interact with Frequency as easily as they would with any traditional Web2 API.

Reviewer: I feel like this isn't very true because of all of the queue systems we have in place. Maybe something closer to "interact with Frequency more like a traditional Web2 API."

Author: I can see that but those are more align with traditional Web2 development.

Key features:
- Modular architecture: Use each service independently to fit your specific needs
- Simplified blockchain interactions: Abstract away complex blockchain operations

> - Web2-like experience: Familiar API patterns for Web2 developers

Reviewer 2: ```suggestion
- Web2-like experience: Familiar REST API patterns for Web2 developers
```

> For a practical implementation example, check out our Social App Template [here](https://github.com/ProjectLibertyLabs/social-app-template), which demonstrates how to integrate these services into a full-fledged application.

Reviewer: Can we bold this or emphasize it?

Author: The link emphasizes this but do you want all of it in bold?

# 📖 Gateway Services <a name="about-project"></a>

Gateway is a collection of services designed to simplify interactions with the Frequency blockchain. It bridges the gap between Web2 and Web3, allowing developers to interact with Frequency as easily as they would with any traditional Web2 API.

Key features:
- Modular architecture: Use each service independently to fit your specific needs
- Simplified blockchain interactions: Abstract away complex blockchain operations
- Web2-like experience: Familiar API patterns for Web2 developers

For a practical implementation example, check out our Social App Template [here](https://github.com/ProjectLibertyLabs/social-app-template), which demonstrates how to integrate these services into a full-fledged application.

<p align="right">(<a href="#-table-of-contents">back to top</a>)</p>

## 🚀 Live Docs <a name="-live-docs"></a>
Visit our Live API Documentation to start exploring the Gateway services.

**[Access Our Live API Docs Here](https://projectlibertylabs.github.io/gateway/)**

> 

Author: extra line

| Content Watcher Service | Monitors and retrieves content updates from the Frequency blockchain. It allows applications to efficiently track and display new content as it's published. | [API Docs](https://projectlibertylabs.github.io/gateway/content-watcher/) | [README](./developer-docs/content-watcher/README.md) |

### Key Features of Gateway Microservices

- **Modularity**: Each service can be used independently, allowing developers to integrate only the functionalities they need.
- **Scalability**: The microservice architecture allows for easy scaling of individual services based on demand.
- **Simplified Blockchain Interaction**: These services abstract complex blockchain operations, providing a familiar API-based interface for developers.
- **Comprehensive Documentation**: Each service comes with detailed API documentation and a README file for easy integration and usage.

### Getting Started with Microservices

> To begin using these microservices:

Author: move down

- [mdBook](https://rust-lang.github.io/mdBook/) for building documentation

### Setup

Clone this repository to your desired folder:

Clone:

2. **Clone the Repository**:
```sh
  git clone git@github.com:ProjectLibertyLabs/gateway.git
  cd gateway
```

### Usage

To run all Gateway services, execute the following command:
3. **Start all Services**

```sh
  ./start.sh
```
This script will start all Gateway microservices using Docker.

To stop the Gateway services, execute the following command:
4. **Stop all Services**

```sh
  ./stop.sh
```
Use this script when you need to stop all running Gateway services.

To build the Gateway Documentation:
5. **Build Gateway Documentation**

```sh
  cd docs
  mdbook build
```

To build and serve the Gateway Documentation:
6. **Build and Serve Gateway Documentation**

```sh
  cd docs
  mdbook serve
```

### Deployment
## Deployment <a name="deployment"></a>

Deployment of the Gateway services have various options. See the [Live Docs](https://projectlibertylabs.github.io/gateway/) for more details.

Deployment of the Gateway documentation occurs via merge to `main` branch via GitHub Actions.

> <p align="right">(<a href="#-table-of-contents">back to top</a>)</p>

Reviewer 3: Do we want to get rid of these? It is a pattern and feels odd to break it just down here.

Author: I had removed all of them, but I can add them back. I thought the page was fairly short to scroll up. I'll add them back.

    S --> GS
    S --> CPS
    S --> CWS
    subgraph Public
        F(Frequency Network)
    end
    AS --> F
    GS --> F
    CPS --> F
    CWS --> F
```

<p align="right">(<a href="#-table-of-contents">back to top</a>)</p>

## Gateway Microservices <a name="gateway-microservices"></a>

Gateway consists of four independent microservices, each designed to handle specific aspects of interaction with the Frequency blockchain. Below is a detailed overview of each service:

| Service | Description | API Documentation | README |
|---------|-------------|-------------------|--------|

> | Account Service | Manages user accounts and authentication on the Frequency blockchain. It handles tasks such as account creation and key management. | [API Docs](https://projectlibertylabs.github.io/gateway/account) | [README](./developer-docs/account/README.md) |

Reviewer: Should we mention "using Sign In With Frequency" with a link to that?

> | Account Service | Manages user accounts and authentication on the Frequency blockchain. It handles tasks such as account creation, and key management. | [API Docs](https://projectlibertylabs.github.io/gateway/account) | [README](./developer-docs/account/README.md) |

Reviewer 2: ```suggestion
| Account Service | Manages user accounts and authentication on the Frequency blockchain. It handles tasks such as account creation and key management. | [API Docs](https://projectlibertylabs.github.io/gateway/account) | [README](./developer-docs/account/README.md) |
```

| Graph Service | Manages social connections and relationships between users on the Frequency network. It handles operations like following/unfollowing users and retrieving social graphs. | [API Docs](https://projectlibertylabs.github.io/gateway/graph/) | [README](./developer-docs/graph/README.md) |

> | Content Publishing Service | Facilitates the creation and publication of content on the Frequency blockchain. It manages tasks such as posting messages, creating threads, and content moderation. | [API Docs](https://projectlibertylabs.github.io/gateway/content-publishing/) | [README](./developer-docs/content-publishing/README.md) |

Reviewer 2: nit: I feel like content moderation may need to be its own service and since we don't really have any of that functionality yet, maybe it doesn't belong here.
```suggestion
| Content Publishing Service | Facilitates the creation and publication of content on the Frequency blockchain. It manages tasks such as posting messages, attachments, replies, and reactions. | [API Docs](https://projectlibertylabs.github.io/gateway/content-publishing/) | [README](./developer-docs/content-publishing/README.md) |
```

> | Content Publishing Service | Facilitates the creation and publication of content on the Frequency blockchain. It manages tasks such as posting messages, creating threads, and content moderation. | [API Docs](https://projectlibertylabs.github.io/gateway/content-publishing/) | [README](./developer-docs/content-publishing/README.md) |

Reviewer: ```suggestion
| Content Publishing Service | Facilitates the publication of content on the Frequency blockchain. It manages tasks such as posting messages, creating threads, and content moderation. | [API Docs](https://projectlibertylabs.github.io/gateway/content-publishing/) | [README](./developer-docs/content-publishing/README.md) |
```

> | Content Watcher Service | Monitors and retrieves content updates from the Frequency blockchain. It allows applications to efficiently track and display new content as it's published. | [API Docs](https://projectlibertylabs.github.io/gateway/content-watcher/) | [README](./developer-docs/content-watcher/README.md) |

Reviewer: ```suggestion
| Content Watcher Service | Monitors and retrieves content updates from the Frequency blockchain. It allows applications to efficiently track new content as it's published. | [API Docs](https://projectlibertylabs.github.io/gateway/content-watcher/) | [README](./developer-docs/content-watcher/README.md) |
```

<p align="right">(<a href="#-table-of-contents">back to top</a>)</p>

### Key Features of Gateway Microservices

- **Modularity**: Each service can be used independently, allowing developers to integrate only the functionalities they need.
- **Scalability**: The microservice architecture allows for easy scaling of individual services based on demand.
- **Simplified Blockchain Interaction**: These services abstract complex blockchain operations, providing a familiar API-based interface for developers.
- **Comprehensive Documentation**: Each service comes with detailed API documentation and a README file for easy integration and usage.

<p align="right">(<a href="#-table-of-contents">back to top</a>)</p>

<!-- GETTING STARTED -->

## 💻 Getting Started <a name="getting-started"></a>

> Gateway offers flexibility in how you can use its services. You can either quickly set up all services or selectively use specific microservices based on your needs. Below, we provide both a Quick Start guide for setting up all services and information on how to get started with individual microservices.

Reviewer: can we link to the quick start?

Author: Yes, we have a link in the table of contents the content is right below this sentence.

Gateway offers flexibility in how you can use its services. You can either quickly set up all services or selectively use specific microservices based on your needs. Below, we provide both a Quick Start guide for setting up all services and information on how to get started with individual microservices.

### 🚀 Quick Start Guide <a name="quick-start-guide"></a>

Follow these steps to quickly get all Gateway services up and running:

1. **Prerequisites**:
In order to run this project you need:
  - [Docker](https://www.docker.com) or Docker compatible layer for running Gateway Services
  - [mdBook](https://rust-lang.github.io/mdBook/) for building documentation

2. **Clone the Repository**:
```sh
  git clone git@github.com:ProjectLibertyLabs/gateway.git
  cd gateway
```

3. **Start all Services**

```sh
  ./start.sh
```
This script will start all Gateway microservices using Docker.

4. **Stop all Services**

```sh
  ./stop.sh
```
Use this script when you need to stop all running Gateway services.

5. **Build Gateway Documentation**

```sh
  cd docs
  mdbook build
```

6. **Build and Serve Gateway Documentation**

```sh
  cd docs
  mdbook serve
```

> <p align="right">(<a href="#-table-of-contents">back to top</a>)</p>

Author: rm
